New UI global search

nfebe opened the pull request:
Resolves: #39162

nfebe committed:
Feat: New UI global search

We are introducing a new search UI that providers a lot more space
for users via a large centralized modal and providers various filters
which can by applied by adding various chips on the UI.

For example, users can now filter their search or scope it by limiting the results
to specific apps, time period and people by apply the appropriate filters on the
new UI, previously filters where applied using text in the search box by prefixing
with `::`.

Resolves: #39162

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

skjnldsv reviewed (commented):
Quick code quality review, (while I know this is a very early state :wink: )
- `console.xxxxx` should be avoided and a logger used instead
- Tabs not spaces
- Please format your classes using the BEM format
- Please also document the various sections in your templates [with](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L36) [html](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L39) [comments](https://github.com/nextcloud/server/blob/16975ae45720945776155f026835cfdaf8491383/apps/files/src/components/FileEntry.vue#L46)

I won't do a review of the code itself as I assume this will heavily change in the future :)

skjnldsv commented on core/src/components/GlobalSearch/SearchFilterChip.vue:
```diff
@@ -0,0 +1,67 @@
+	},
+	methods: {
+		removeChip() {
+			this.$emit('remove-filter', this.filter)
```
It's easier to use standard event names like `delete`. You don't really need to specify that this is about filters as we'll see the event on the `<FilterChip />` component in template :)

Fenn-CS commented on core/src/views/GlobalSearchModal.vue:
```diff
@@ -0,0 +1,399 @@
+		})
+	},
+	methods: {
+		find(query) {
+			if (query.length) {
+				const newResults = []
+				let activeSearchProviders = []
+				if (this.filteredProviders.length > 0) {
+					activeSearchProviders = this.filteredProviders
+				} else {
+					activeSearchProviders = this.providers
+				}
+				activeSearchProviders.forEach(provider => {
+					console.debug('Search type', provider.id)
+					const request = globalSearch({ type: provider.id, query, cursor: null }).request
+					request().then((response) => {
+						newResults.push({ id: provider.id, provider: provider.name, results: response.data.ocs.data.entries })
+						console.debug('New results', newResults)
+						console.debug('Global search results:', newResults)
+						this.updateResults(newResults)
+					})
+				})
+			} else {
+				this.results = []
+			}
+		},
```
@artonge Please see if there are any performance concerns here... The method is debounced before call.

ChristophWurst commented on core/src/views/GlobalSearch.vue:
```diff
@@ -0,0 +1,67 @@
+  -->
+<template>
+	<div>
+		<NcButton aria-label="Global search" @click="toggleGlobalSearch">
```
l10n

nfebe committed:
Add unified search toggle in config

The new global search UI is new and might be unstable, hence
we are giving users the option to use the old unified search UI, if
the encounter signficant bottlenecks.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>
Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

Fenn-CS commented:
So I spoke with @AndyScherzinger and we going to merge this, but keep the old search as the default. This would allow for the team to be able to see the new ui on tech preview by setting `unified-search.enabled => false` in the config.

Next week, we would have to send in updates to fix these by priority

- [ ] Icon inversion which does not work well now since icons come in white or black instead of one consistent color.
- [ ] Use the new people filter that @marcoambrosini is working (Current people filter not completed)
- [ ] Update back-end API (if ability not already present) to show "Search in App" dynamically.

Fenn-CS commented:
/compile amend /

ChristophWurst commented:
@Fenn-CS could you please tell us how to test this? The UI looks identical to `master` for me

![image](https://github.com/nextcloud/server/assets/1374172/426472eb-dc14-448d-8763-082c5847f507)

Fenn-CS commented:
@ChristophWurst set `unified-search.enabled => false` in the config file

ChristophWurst commented:
> @ChristophWurst set `unified-search.enabled => false` in the config file

No change and the config key doesn't exist in the code. Do I need any external apps for this feature?

ChristophWurst commented:
State is ok for a very bumpy first version hidden behind a feature flag :+1: 

The biggest issue is white icons and text on white background. My theme is set to #C935BA, so buttons are white. The situation might be better with the default color.

- [ ] BUG: Clicking outside the search modal needs two clicks on the magnifier icon to get the modal back.
- [ ] BUG: Some app icons are white on white.
- [ ] BUG: Modified timestamp bubbles change size on hover, have no pointing cursor.
- [ ] BUG: Generally no feedback for *no results*.
- [ ] BUG: *Search people* has white text on white background.
- [ ] BUG: *Search people* has a tiny user bubble, barely wider than the avatar.
- [ ] BUG: *Search people* *No result* goes to the top, overlaps other input.

Fenn-CS commented:
/compile amend /

Fenn-CS commented:
Cypress tests keep timing out but are unrelated. Check a couple of tickets with same issues...

AndyScherzinger commented:
Not up to date with main branch. I guess it needs another rebase and compile 🤔

blizzz commented:
> Not up to date with main branch. I guess it needs another rebase and compile 🤔

No, there are no conflicts on the compiled assets

blizzz commented:
CI complains however:

```
ERROR: There is 1 use of triple dot instead of ellipsis:
At line 17 in file /drone/src/build/../core/src/views/GlobalSearchModal.vue
```

Does it refer to https://github.com/nextcloud/server/pull/40823/files#diff-72e59d550e6509240d3c7bad01689aa5d0f9ce1c3b530fb7a45100fedebf37ffR18 ? Should be replaced with … .

Fenn-CS commented:
/compile amend /

marcoambrosini committed:
Create searchable list component

Signed-off-by: Marco <marcoambrosini@icloud.com>

marcoambrosini commented:
@Fenn-CS I pushed the searchable list component for people. To add it to global search modal copy-paste this:

```
<!-- People search popover -->
    <SearchableList :label-text="t('core', 'Search people')"
	    :search-list="peopleSeclectProps.options"
	    :empty-content-text="t('core', 'Not found')">
	    <template #trigger>
		    <NcButton>
			    {{ t('core', 'People') }}
		    </NcButton>
	    </template>
    </SearchableList>
```

With a few tweaks, we should be able to easily adapt this for the apps too

AndyScherzinger commented:
/compile amend /

nfebe committed:
Use new SearchableList component for people filter

This commit migrates away from NcSelect which has a couple of
accesibility and display problems currently, hence a new component
`SearchableList` is now used.

Signed-off-by: fenn-cs <fenn25.fn@gmail.com>

Fenn-CS commented:
/compile amend /

Fenn-CS commented:
/compile /

nextcloud-command committed:
chore(assets): Recompile assets

Signed-off-by: nextcloud-command <nextcloud-command@users.noreply.github.com>

Fenn-CS commented:
Bringing this down here,  so it remains visible and easily transferable to a new ticket.

TODO

- [ ] Icon inversion which does not work well now since icons come in white or black instead of one consistent color. (Black icons turn white and white icons turn black, leading to to white on white background issues)
- [ ] Use the new people filter (selector) that @marcoambrosini is working on (Current people filter not completed) - DONE, however, `NcPopover` which is used has issues with closing triggers. 
- [ ] Update back-end API (if ability not already present) to show "Search in App" dynamically.

ChristophWurst commented:
And all of https://github.com/nextcloud/server/pull/40823#issuecomment-1804326745

Fenn-CS commented:
Thanks @ChristophWurst updated.
